lambda/lambda.ml

```diff
@@ -46,9 +40,62 @@ type field_read_semantics =
   | Reads_agree
   | Reads_vary
 
-type alloc_mode =
-  | Alloc_heap
-  | Alloc_local
+include (struct
+
+  type alloc_mode =
+    | Alloc_heap
+    | Alloc_local
+
+  let alloc_heap = Alloc_heap
+
+  let alloc_local : alloc_mode =
+    if Config.stack_allocation then Alloc_local
+    else Alloc_heap
+
+  let join_mode a b =
+    match a, b with
+    | Alloc_local, _ | _, Alloc_local -> Alloc_local
+    | Alloc_heap, Alloc_heap -> Alloc_heap
```

Comment:
If you like, I think you can put `join_mode` on the other side of the airtight hatchway by writing it instead as something like:
```ocaml
let join_mode a b =
  match a, b with
  | (Alloc_local as x), _
  | _, (Alloc_local as x) -> x
  | Alloc_heap, (Alloc_heap as x) -> x
```